### PR TITLE
add Bing Search Engine Subdomains Collector

### PR DESCRIPTION
**Pull request by nixawk**
```
msf > use auxiliary/gather/bing_subdomains_collector 
msf auxiliary(bing_subdomains_collector) > set DOMAIN rapid7.com
DOMAIN => rapid7.com
msf auxiliary(bing_subdomains_collector) > run 

[*] Searching Bing for subdomains from rapid7.com
[+] rapid7.com subdomain: www.rapid7.com - 54.230.150.109
[+] rapid7.com subdomain: insight.rapid7.com - 52.71.140.111,52.72.3.42,54.165.221.90
[+] rapid7.com subdomain: upnp-check.rapid7.com - 54.214.49.70
[+] rapid7.com subdomain: information.rapid7.com - 199.15.214.152
[+] rapid7.com subdomain: experttracks.rapid7.com - 54.86.70.125
[+] rapid7.com subdomain: download2.rapid7.com - 23.62.109.40,23.62.109.8
[+] rapid7.com subdomain: vdb-web.labs.rapid7.com - 54.214.49.70
[*] Auxiliary module execution completed
msf auxiliary(bing_subdomains_collector) > notes 
[*] Time: 2016-02-17 03:04:06 UTC Note: host=54.240.168.12 type=Bing Search Subdomains data={"www.rapid7.com"=>["54.230.150.109"], "insight.rapid7.com"=>["52.71.140.111", "52.72.3.42", "54.165.221.90"], "upnp-check.rapid7.com"=>["54.214.49.70"], "information.rapid7.com"=>["199.15.214.152"], "experttracks.rapid7.com"=>["54.86.70.125"], "download2.rapid7.com"=>["23.62.109.40", "23.62.109.8"], "vdb-web.labs.rapid7.com"=>["54.214.49.70"]}
```

### CODE REVIEW

**Comment by wvu-r7** [comment]
Something with ```ip:``` would be neat, too.

File: modules/auxiliary/gather/bing_subdomains_collector.rb [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+        'rport' => rport_bing,
+        'vhost' => rhost_bing,
+        'method' => 'GET',
+        'uri' => '/search',
+        'vars_get' => {
+          'first' => num,
+          'q' => "domain:#{domain}"
+        })
+
+      if resp && resp.redirect?
```
Comment (wvu-r7) [inline modules/auxiliary/gather/bing_subdomains_collector.rb]: You should be able to use ```send_request_cgi!``` above.

Comment (all3g) [inline modules/auxiliary/gather/bing_subdomains_collector.rb]: Thanks @wvu-r7, ```send_request_cgi!``` (note the !) does follow redirects.

**Comment by all3g** [comment]
Thanks @wvu-r7. domains/ips are both in print_* / notes. Please see the module demo.

**Comment by wvu-r7** [comment]
@all3g: Thanks for the changes. I meant the ```ip:``` query in Bing, actually. :)

**Comment by all3g** [comment]
@wvu-r7 A ```IP_SEARCH``` switch can be given to users, and they can choose how to gather subdomains.

```
OptBool.new('IP_SEARCH', [ false, "The ip of domain is used to locate subdomains for", true])
```

**Comment by all3g** [comment]
This module use bing dorks as follow:

- [x] ```domain:demo.com```
- [x] ```ip:xxx.xxx.xxx.xxx```

----


- **TARGET** option can be a domain or ip address.
- **IP_SEARCH** option lets us gather subdomains with the ip of subdomains. If disable, bing will only search subdomains with domain.

## **Search subdomains with domain**

```
msf > use auxiliary/gather/bing_subdomains_collector 
msf auxiliary(bing_subdomains_collector) > show options 

Module options (auxiliary/gather/bing_subdomains_collector):

   Name       Current Setting  Required  Description
   ----       ---------------  --------  -----------
   IP_SEARCH  true             no        Enable ip of subdomains to locate subdomains
   TARGET                      yes       The target to locate subdomains for, ex: rapid7.com, 8.8.8.8

msf auxiliary(bing_subdomains_collector) > set TARGET exploit-db.com
TARGET => exploit-db.com
msf auxiliary(bing_subdomains_collector) > run 

[*] Searching Bing for subdomains from exploit-db.com
[+] domain:exploit-db.com subdomain: www.exploit-db.com - 192.124.249.8
[*] Searching Bing for subdomains from 192.124.249.8
[+] ip:192.124.249.8 suddomain: www.exploit-db.com
[+] ip:192.124.249.8 suddomain: www.imvbox.com
[+] ip:192.124.249.8 suddomain: www.tuckerssolicitors.com
[+] ip:192.124.249.8 suddomain: www.city-academy.com
[*] Auxiliary module execution completed
```

## **Search subdomains with ip**

```
msf auxiliary(bing_subdomains_collector) > set TARGET 192.124.249.8
TARGET => 192.124.249.8
msf auxiliary(bing_subdomains_collector) > run 

[*] Searching Bing for subdomains from 192.124.249.8
[+] ip:192.124.249.8 suddomain: www.exploit-db.com
[+] ip:192.124.249.8 suddomain: www.imvbox.com
[+] ip:192.124.249.8 suddomain: www.tuckerssolicitors.com
[+] ip:192.124.249.8 suddomain: www.city-academy.com
[+] ip:192.124.249.8 suddomain: cardexchangesolutions.com
[*] Auxiliary module execution completed
```

----

Could any body show me how to save both of domain and ip in **```bing_search_ip```** ? **```report_host```** can't save all domains with same ip.

**Comment by wvu-r7** [comment]
@all3g: You are way too cool!

File: modules/auxiliary/gather/bing_subdomains_collector.rb [hunk 2]
```diff
@@ -0,0 +1,145 @@
...
+
+    results = bing_search(dork)
+    results.each do |subdomain|
+      next if domains.include?(subdomain)
+      next unless valid?(ip, subdomain)
+      report_host(host: ip, name: subdomain)
+      print_good("#{dork} suddomain: #{subdomain}")
+      domains[subdomain] = ip
+    end
+    return unless domains
```
Comment (jlee-r7) [inline modules/auxiliary/gather/bing_subdomains_collector.rb]: domains will always be a true value here. Maybe you meant `domains.empty?`

File: modules/auxiliary/gather/bing_subdomains_collector.rb [hunk 3]
```diff
@@ -0,0 +1,145 @@
...
+      ips = Rex::Socket.getaddresses(subdomain)
+      ips.each do |ip|
+        report_host(host: ip, name: subdomain)
+        print_good("#{dork} subdomain: #{subdomain} - #{ip}")
+        bing_search_ip(ip) if datastore['IP_SEARCH']
+      end
+      domains[subdomain] = ips
+    end
+
+    return unless domains
```
Comment (jlee-r7) [inline modules/auxiliary/gather/bing_subdomains_collector.rb]: domains will always be a true value here. Maybe you meant `domains.empty?`

**Comment by all3g** [comment]
Thanks @jlee-r7 .

File: modules/auxiliary/gather/bing_subdomains_collector.rb [hunk 4]
```diff
@@ -0,0 +1,145 @@
...
+      'Description' => %q(
+        This module can be used to gather subdomains about a domain
+        from Bing Search Results.
+      ),
+      'Author' => [ 'Nixawk' ],
+      'License' => MSF_LICENSE))
+
+    register_options(
+      [
+        OptString.new('TARGET', [ true, "The target to locate subdomains for, ex: rapid7.com, 8.8.8.8"]),
```
Comment (jlee-r7) [inline modules/auxiliary/gather/bing_subdomains_collector.rb]: `TARGET` has specific a meaning in other parts of the framework. Maybe `DOMAIN_OR_IP`? Not sure what the best thing would be here.

**Comment by jlee-r7** [comment]
Now that I look at this again, `bing_search_domain` and `bing_search_ip` are almost identical. Can you pull the common stuff out into another method?

**Comment by all3g** [comment]
Yeah, I'll merge yahoo/bing/... into a module. 
Close the pull now, and discuss it [here](https://github.com/rapid7/metasploit-framework/pull/6576).

**Pull request closed without merging**